### PR TITLE
Fix gateway service dependencies

### PR DESCRIPTION
**Pull request by dddddai**
**Please provide a description of this PR:**
Fixes https://github.com/istio/istio/issues/42400
The root cause is https://github.com/istio/istio/pull/39168, we should check the **old service visibility** as well, otherwise users can not opt in

**To help us figure out who should review this PR, please put an X in all the areas that this PR affects.**

- [ ] Ambient
- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Installation
- [x] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

**Please check any characteristics that apply to this pull request.**

- [ ] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

File: pilot/pkg/xds/proxy_dependencies.go [hunk 1]
```diff
@@ -73,11 +73,13 @@ func checkProxyDependencies(proxy *model.Proxy, config model.ConfigKey, push *mo
...
 			// If config is ServiceEntry, name of the config is service's FQDN
-			svc, exist := push.ServiceIndex.HostnameAndNamespace[host.Name(config.Name)][config.Namespace]
-			if exist {
-				if !push.IsServiceVisible(svc, proxy.Metadata.Namespace) {
-					return false
-				}
+			hostname := host.Name(config.Name)
+			// gateways have default sidecar scopes
+			if proxy.SidecarScope.GetService(hostname) == nil &&
+				proxy.PrevSidecarScope.GetService(hostname) == nil {
```
Comment (dddddai) [inline pilot/pkg/xds/proxy_dependencies.go:79]: BTW I think we could expand this to other configs(VS, DR) ? just like what we do for sidecars

**Review by hzxuzhonghu — COMMENTED** [review]
I do not get the fix. The issue is cluster shouldn't be visible, We get services from SidecarScope.

While this patch only check proxy needs push or not, it does not influence SidecatScope

File: pilot/pkg/xds/proxy_dependencies.go (hunk 1 above)
Comment (hzxuzhonghu) [inline pilot/pkg/xds/proxy_dependencies.go:79]: proxy.PrevSidecarScope can be nil

Comment (dddddai) [reply pilot/pkg/xds/proxy_dependencies.go:79]: I've already checked it
https://github.com/istio/istio/pull/42422/files#diff-42d26d46fa9490c70f6bd91080addf3ecc57aa44ebae320708ae58c824af8a9dR553

**Comment by dddddai** [comment]
> The issue is cluster shouldn't be visible, We get services from SidecarScope.

The issue is we are not sending new clusters to the gateway, because it does not get xds push if [the updated service is not visible](https://github.com/istio/istio/blob/1.16.1/pilot/pkg/xds/proxy_dependencies.go#L78), while we should push since the service visibility is changed

File: pilot/pkg/xds/proxy_dependencies.go (hunk 1 above)
Comment (hzxuzhonghu) [reply pilot/pkg/xds/proxy_dependencies.go:79]: lg

File: releasenotes/notes/42400.yaml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+apiVersion: release-notes/v2
+kind: bug-fix
+area: traffic-management
+issue:
+- 42400
+releaseNotes:
+- |
+  **Fixed** updating service visibility does not push XDS to gateways
```
Comment (ericvn) [inline releasenotes/notes/42400.yaml:8]: Nit: Add a period at the end of the sentence:
```suggestion
  **Fixed** updating service visibility does not push XDS to gateways.
```

**Review by howardjohn — APPROVED** [review]

**Pull request merged**